File: Android/app/src/main/java/fr/uge/structsure/bluetooth/cs108/Connexion.java
package fr.uge.structsure.bluetooth.cs108;

import android.content.Context;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;

import fr.uge.structsure.MainActivity;

import com.csl.cslibrary4a.ReaderDevice;

import java.util.ArrayList;

public class Connexion {
    private DeviceScanTask deviceScanTask;
    private DeviceConnectTask deviceConnectTask;
    public final ArrayList<ReaderDevice> readersList = new ArrayList<>();
    private final Handler mHandler = new Handler();
    private Context context = null;

    public Connexion(Context context) {
        this.context = context;
        if (!MainActivity.csLibrary4A.isBleConnected()) readersList.clear();
        checkRunnable.run();
    }

    public void onItemClick(ReaderDevice readerDevice) {
//        ReaderDevice readerDevice = readersList.get(position);    // Un ReaderDevice est l'objet qui représente un périphérique BLE compatible

        if (MainActivity.csLibrary4A.isBleConnected() && readerDevice.isConnected() && (readerDevice.getSelected())) {
            // If THIS device is already connected, disconnect the device
            System.out.println("[CONNECT] - Disconnecting from device " + readerDevice.getName());
            MainActivity.csLibrary4A.disconnect(false);
            readersList.clear();
        } else if (!MainActivity.csLibrary4A.isBleConnected() && !readerDevice.getSelected()) {
            // If not connected yet to the BLE device
            boolean validStart = deviceConnectTask == null || deviceConnectTask.getStatus() == AsyncTask.Status.FINISHED;
            if (validStart) {    // Creates a new DeviceConnectTask if not started yet
                if (deviceScanTask != null) deviceScanTask.cancel(true);    // Stops the scan task?
                System.out.println("[CONNECT] - Connecting to device " + readerDevice.getName());
                deviceConnectTask = new DeviceConnectTask(readerDevice);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
                    deviceConnectTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
                } else {
                    deviceConnectTask.execute();
                }
            }
        }

//                if (readersList.size() > position) {    // Update displayed list
//                  readerDevice.setSelected(!readerDevice.getSelected());    // Invert the selected state of the device
//                    readersList.set(position, readerDevice);
//                    for (int i = 0; i < readersList.size(); i++) {
//                        if (i != position) {
//                            ReaderDevice readerDevice1 = readersList.get(i);
//                            if (readerDevice1.getSelected()) {
//                                readerDevice1.setSelected(false);
//                                readersList.set(i, readerDevice1);
//                            }
//                        }
//                    }
//                }
//                readerListAdapter.notifyDataSetChanged();
    }

    /**
     * Once connected
     */
    public void onStop() {
        mHandler.removeCallbacks(checkRunnable);
        if (deviceScanTask != null) {
            deviceScanTask.cancel(true);
        }
        if (deviceConnectTask != null) {
            deviceConnectTask.cancel(true);
        }
    }

    private final Runnable checkRunnable = new Runnable() {
        @Override
        public void run() {
            boolean operating = MainActivity.csLibrary4A.isBleConnected();
            if (!operating && deviceScanTask != null) {
                if (!deviceScanTask.isCancelled())   operating = true;
            }
            if (!operating && deviceConnectTask != null) {
                if (!deviceConnectTask.isCancelled())   operating = true;
            }
            if (!operating) {
                deviceScanTask = new DeviceScanTask(context, readersList, r -> onItemClick(r));
                deviceScanTask.execute();
            }
            mHandler.postDelayed(checkRunnable, 5000);
        }
    };
}

File: Android/app/src/main/java/fr/uge/structsure/bluetooth/cs108/Connexion.kt
package fr.uge.structsure.bluetooth.cs108

import android.content.Context
import android.os.AsyncTask
import android.os.Build
import androidx.compose.runtime.mutableStateListOf
import com.csl.cslibrary4a.ReaderDevice
import fr.uge.structsure.MainActivity

class Connexion(private val context: Context) {
    private var scanTask: ScanTask? = null
    private var connectTask: ConnectTask? = null
    var readersList: MutableList<ReaderDevice> = mutableStateListOf()

    init {
        if (!MainActivity.csLibrary4A.isBleConnected) readersList.clear()
    }
    
    fun startScan() {
        val scanning = scanTask != null && !scanTask!!.isCancelled

        if (!scanning) {
            println("Starting scan")
            scanTask = ScanTask(context, readersList) {
                    r: ReaderDevice -> onItemClick(r)
            }
            scanTask!!.execute()
        }
    }

    fun onItemClick(readerDevice: ReaderDevice) {
        println("Reader clicked: $readerDevice")
        if (MainActivity.csLibrary4A.isBleConnected && readerDevice.isConnected && (readerDevice.selected)) {
            // If THIS device is already connected, disconnect the device
            println("[CONNECT] - Disconnecting from device " + readerDevice.name)
            MainActivity.csLibrary4A.disconnect(false)
            readersList.clear()
        } else if (!MainActivity.csLibrary4A.isBleConnected && !readerDevice.selected) {
            // If not connected yet to the BLE device
            val validStart =
                connectTask == null || connectTask!!.status == AsyncTask.Status.FINISHED
            if (validStart) {    // Creates a new DeviceConnectTask if not started yet
                if (scanTask != null) scanTask!!.cancel(true) // Stops the scan task?

                println("[CONNECT] - Connecting to device " + readerDevice.name)
                connectTask = ConnectTask(readerDevice) {r -> afterConnection(r)}
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
                    connectTask!!.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
                } else {
                    connectTask!!.execute()
                }
            }
        }
    }

    private fun afterConnection(device: ReaderDevice?) {
        if (device != null) {
            readersList.remove(device)
            readersList.add(device)

//            val inventoryRfidTask = RfidTask(
//                context, -1, -1, 0, 0, 0, 0,
//                false
//            )
//            RfidTask(context).execute()
//            InventoryRfidTask.startScan(context)
//             RfidTask.startScan(context)
//            MainActivity.csLibrary4A.setTagRead(1)
//            MainActivity.csLibrary4A.setAccessBank(3)
//            MainActivity.csLibrary4A.setAccessOffset(112)
//            MainActivity.csLibrary4A.setAccessCount(1)
//            MainActivity.csLibrary4A.setNotificationListener {
//                MainActivity.csLibrary4A.appendToLog("setNotificationListener TRIGGER key is pressed in search inventory.")
//            }
//            MainActivity.csLibrary4A.startOperation(RfidReaderChipData.OperationTypes.TAG_SEARCHING)
//            var geigerSearchTask = InventoryRfidTask(
//                context, -1, -1, 0, 0, 0, 0, true, true,
//                null, null, null /* textView */, null,
//                null, null, null, null, null, null
//            )
//            geigerSearchTask.execute()

        }

    }

    /**
     * Once connected
     */
    fun onStop() {
        if (scanTask != null) {
            println("Scan stopped")
            scanTask!!.cancel(true)
        }
        if (connectTask != null) {
            connectTask!!.cancel(true)
        }
    }
}